### PR TITLE
avoid permission problems of tempdir on shared servers

### PR DESCRIPTION
**Pull request by lzutao**
### Description/Motivation/Screenshots ###
Avoid permission problems on shared server with multiple users

### How Has This Been Tested? ###

| Architecture |          Yes/No          | Comments                                  |
| ------------ | :----------------------: | ----------------------------------------- |
| x86-32       | :heavy_multiplication_x: | Replace with :heavy_check_mark: if tested |
| x86-64       | :heavy_check_mark: |                                           |
| ARM          | :heavy_multiplication_x: |                                           |
| AARCH64      | :heavy_multiplication_x: |                                           |
| MIPS         | :heavy_multiplication_x: |                                           |
| POWERPC      | :heavy_multiplication_x: |                                           |
| SPARC        | :heavy_multiplication_x: |                                           |
| RISC-V       | :heavy_multiplication_x: |                                           |
| `make tests` | :heavy_multiplication_x: |                                           |

### Checklist ###

- [x] My PR was done against the `dev` branch, not `master`.
- [ ] My code follows the code style of this project.
- [ ] My change includes a change to the documentation, if required.
- [ ] My change adds tests as appropriate.
- [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by Grazfather** [comment]
.cache directory isn't guaranteed to exist.

**Comment by lzutao** [comment]
It will be created if not by the script.

File: gef.py [hunk 1]
```diff
@@ -93,16 +93,7 @@
...
-        """Replace range() builtin with an iterator version."""
-        if len(args) < 1:
-            raise TypeError()
-        start, end, step = 0, args[0], 1
-        if len(args) == 2: start, end = args
-        if len(args) == 3: start, end, step = args
-        for n in itertools.count(start=start, step=step):
-            if (step>0 and n >= end) or (step<0 and n<=end): break
-            yield n
+    range = xrange
```
Comment (hugsy) [inline gef.py]: This won't be accepted, reason: see a9949e896d4cf787aa56f1e899163ec8e42704c3

File: gef.py [hunk 2]
```diff
@@ -206,15 +210,15 @@ def update_gef(argv):
 __gef_int_stream_buffer__              = None
 __gef_redirect_output_fd__             = None
 
+HOME_DIR = os.getenv("HOME")
 DEFAULT_PAGE_ALIGN_SHIFT               = 12
 DEFAULT_PAGE_SIZE                      = 1 << DEFAULT_PAGE_ALIGN_SHIFT
-GEF_RC                                 = os.path.join(os.getenv("HOME"), ".gef.rc")
-GEF_TEMP_DIR                           = os.path.join(tempfile.gettempdir(), "gef")
+GEF_RC                                 = os.path.join(HOME_DIR, ".gef.rc")
+GEF_TEMP_DIR                           = os.path.join(HOME_DIR, ".cache", "gef")
```
Comment (hugsy) [inline gef.py]: I'd much prefer something like:
```python
os.path.join(tempfile.gettempdir(), os.getenv("USER"), "gef")
```

`gettempdir()` has the advantage of being deleted every reboot. Why use `~/.cache` in particular?

File: gef.py [hunk 3]
```diff
@@ -206,15 +211,15 @@ def update_gef(argv):
...
 GEF_MAX_STRING_LENGTH                  = 50
 
 GDB_MIN_VERSION                         = (7, 7)
-GDB_VERSION_MAJOR, GDB_VERSION_MINOR    = [int(_) for _ in re.search(r"(\d+)[^\d]+(\d+)", gdb.VERSION).groups()]
-GDB_VERSION                             = (GDB_VERSION_MAJOR, GDB_VERSION_MINOR)
+GDB_VERSION                             = tuple(map(int, gdb.VERSION.split('.')[:3]))
```
Comment (hugsy) [inline gef.py]: This won't be accepted either, see b060b2807383f43169ad99c35482ce533f343cf9

File: gef.py [hunk 4]
```diff
@@ -4789,7 +4794,7 @@ def create_or_edit_structure(self, mod_name, struct_name):
         else:
             info("Editing '{:s}'".format(fullname))
 
-        cmd = os.getenv("EDITOR").split() if os.getenv("EDITOR") else ["nano",]
+        cmd = (os.getenv("EDITOR") or "nano").split()
```
Comment (hugsy) [inline gef.py]: Although this can be accepted, as it is cleaner, it is unrelated to this PR, therefore making it hard for us to track. Please make a different PR.

File: gef.py [hunk 5]
```diff
@@ -95,17 +96,21 @@
...
     def range(*args):
         """Replace range() builtin with an iterator version."""
-        if len(args) < 1:
-            raise TypeError()
-        start, end, step = 0, args[0], 1
-        if len(args) == 2: start, end = args
-        if len(args) == 3: start, end, step = args
-        for n in itertools.count(start=start, step=step):
-            if (step>0 and n >= end) or (step<0 and n<=end): break
+        if len(args) == 0:
```
Comment (hugsy) [inline gef.py]: This is really confusing, does it really give a gain in performance?
Also this is unrelated to this PR. If you want to re-write the `range()` function, please make a separate PR.

File: gef.py (hunk 2 above)
Comment (lzutao) [reply gef.py]: What if an user on the same shared server squat `/tmp/<user>` of other users?

**Comment by lzutao** [comment]
I removed unrelated change.

File: gef.py (hunk 2 above)
Comment (hugsy) [reply gef.py]: You will have the sticky bit to protect your stuff:
```
hugsy@ubuntu-dev ~ $ mkdir /tmp/hugsy
hugsy@ubuntu-dev ~ $ touch /tmp/hugsy/foo
hugsy@ubuntu-dev ~ $ sudo -u user touch /tmp/hugsy/bar
touch: cannot touch '/tmp/hugsy/bar': Permission denied
```

Comment (lzutao) [reply gef.py]: Placing stuff in `~/.cache` avoid that problems. 
Most browsers store their cache, and some software store video/picture thumbnails there.

**Comment by lzutao** [comment]
Ping.

**Comment by hugsy** [comment]
After some talks, we've come to agree that `~/.cache` is not a good candidate for storing user temporary content. The idea is that the content in `GEF_TEMP_DIR` is disposable, and `~/.cache` is never emptied (except if explicitly `rm`-ed by the user).

You did raise a good point related to the squatting in `/tmp`. A good way mitigate that would be to add some random value generated when gef starts like `/tmp/<random_string>`. This also would pose a problem for people who use gef a lot for remote sessions, as they will constantly have to re-download the content they already have (since a new /tmp location will be created every time they start gef).

What I would prefer is to simply let the user decide of the location via a `gef config` setting: leave `GEF_TEMP_DIR` assigned to `/tmp/gef` by default, but load the global setting in case like yours where people do not to use that location.

Feel free to update this PR if you want to implement that, or I can take care of this soon.

Thanks.

**Comment by lzutao** [comment]
I don't believe I could make it better than you.
So I leave it to you.

**Comment by hugsy** [comment]
> I don't believe I could make it better than you.
> So I leave it to you.

Can you open an issue so we don't lose track of what's been done and discussed?

**Comment by lzutao** [comment]
Sure, done in #486
